File: src/org/kobjects/nlp/api/Definition.java
package org.kobjects.nlp.api;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.kobjects.nlp.latin.Conjugator;

public class Definition {
	public final WordType type;
	public Map<Language, List<Definition>> translations = new LinkedHashMap<>();
	public Gender genus;
	public String description;
	public Map<Form, Word> forms = new LinkedHashMap<>();
	
	public Definition(WordType type, String description) {
		this.type = type;
		this.description = description;
	}
	
	@Override
	public String toString() {
		return type + " " + description;
	}
	
	public void addFormless(String word) {
		forms.put(Form.EMPTY, new Word(word, Form.EMPTY, this));
	}
	
	public void addForms(Map<Form, String> map) {
		for (Map.Entry<Form, String> entry : map.entrySet()) {
			forms.put(entry.getKey(), new Word(entry.getValue(), entry.getKey(), this));
		}
	}
	
}

File: src/api/Definition.js
import Form from "./Form.js";
import FormBuilder from "./FormBuilder.js";
import Word from "./Word.js";

export default class Definition {
	constructor(type, description, codes) {
		this.type = type;
		this.description = description;
		this.translations = new Map();
		this.forms = new Map();
		this.genus = undefined;
		this.codes = codes;
	}

	// Acts like how this.forms.get(identicalOtherFormInstance) would've worked
	getWord(form) {
		const match = Array.from(this.forms).find(([formWordKey]) => (
			formWordKey.equals(form)
		));
		return match && match[1];
	}
	
	toString() {
		return this.type + " " + this.description;
	}
	
	addFormless(word) {
		this.forms.set(FormBuilder.EMPTY, new Word(word, FormBuilder.EMPTY, this));
		return this;
	}
	
	addForms(map) {
		map.forEach((wordString, form) => {
			this.forms.set(form, new Word(wordString, form, this));
		});
		return this;
	}
}
